dhis-2/dhis-api/src/main/java/org/hisp/dhis/icon/IconService.java

```diff
@@ -40,17 +40,10 @@ public interface IconService {
   /**
    * Gets data about all the icons in the system
    *
+   * @param iconCriteria criteria to fetch icons
    * @return a collection of data about all the icons in the system
    */
-  List<Icon> getIcons();
-
-  /**
-   * Gets icons tagged with all given keywords.
-   *
-   * @param keywords collection of keywords
-   * @return a collection of matching icons
-   */
-  List<Icon> getIcons(String[] keywords);
+  List<? extends Icon> getIcons(IconCriteria iconCriteria);
 
   /**
    * Gets the icon associated to a key, if it exists
```

Comment:
I don't know all of it, but if it implements an interface, it should return the `Icon` in the first place to avoid IDE warnings and make the code cleaner.

Reply:
Yes. I did try that and surprising for me that IDE started complaining. So I then had to use this

Reply:
I think the best way is to change the signature in the interface:

`
List<Icon> getIcons(IconCriteria iconCriteria);
`

then in the `getIcons()` method in `DefaultIconService` you can return:

`
return new ArrayList<>(customIconStore.getIconsByKeywords(iconCriteria.getKeywords()));
`

Reply:
You should then be able to use `switch` :

```
  @Override
  @Transactional(readOnly = true)
  public List<Icon> getIcons(IconCriteria iconCriteria) {
    switch (iconCriteria.getType()) {
      case DEFAULT -> {
        return defaultIcons.values().stream()
            .filter(icon -> Set.of(icon.getKeywords()).containsAll(iconCriteria.getKeywords()))
            .toList();
      }
      case CUSTOM -> {
        return new ArrayList<>(customIconStore.getIconsByKeywords(iconCriteria.getKeywords()));
      }
      default -> {
        return Stream.concat(
                defaultIcons.values().stream()
                    .filter(
                        icon -> Set.of(icon.getKeywords()).containsAll(iconCriteria.getKeywords()))
                    .toList()
                    .stream(),
                customIconStore.getIconsByKeywords(iconCriteria.getKeywords()).stream())
            .toList();
      }
    }
  }
```

Reply:
That's why i have mentioned it [here](https://github.com/dhis2/dhis2-core/pull/15929#discussion_r1450861309.).
So either we have to fix the code or the sonarcloud configurations.

<img width="736" alt="Screenshot 2024-01-15 at 09 17 22" src="https://github.com/dhis2/dhis2-core/assets/13900328/694cfbf9-b3e8-4444-889e-97d217581de9">

Reply:
ok but that is about mutability and should be fixable, I'll have a look

Reply:
```
  private final Map<String, Icon> defaultIcons =
      Arrays.stream(DefaultIcon.Icons.values())
          .map(DefaultIcon.Icons::getVariants)
          .flatMap(Collection::stream)
          .collect(Collectors.toMap(DefaultIcon::getKey, Function.identity()));

  @Override
  @Transactional(readOnly = true)
  public List<Icon> getIcons(IconCriteria iconCriteria) {
    switch (iconCriteria.getType()) {
      case DEFAULT -> {
        return defaultIcons.values().stream()
            .filter(icon -> Set.ofli(icon.getKeywords()).containsAll(iconCriteria.getKeywords()))
            .toList();
      }
      case CUSTOM -> {
        return new ArrayList<>(customIconStore.getIconsByKeywords(iconCriteria.getKeywords()));
      }
      default -> {
        return Stream.concat(
                defaultIcons.values().stream()
                    .filter(
                        icon -> Set.of(icon.getKeywords()).containsAll(iconCriteria.getKeywords()))
                    .toList()
                    .stream(),
                customIconStore.getIconsByKeywords(iconCriteria.getKeywords()).stream())
            .toList();
      }
    }
  }
```
This should work, but you must roll back the `stream` change and make the `getIconsByKeywords` return a List<CustomIcon>. The toList() enforced in the latest version of Java returns the implementation rather than the abstraction, so it is a bit unfortunate.